py4DSTEM/process/latticevectors/index.py

```diff
@@ -53,3 +53,158 @@ def index_bragg_directions(x0, y0, ux, uy, vx, vy, bragg_x, bragg_y):
+    ideal_lattice.remove_points(deletePeaks)
+
+    return ideal_lattice
+
+def index_point_list_array(pointlistarray, lattice, maxPeakSpacing, mask=None):
+    """
+    Indexes the Bragg peaks in a PointListArray based on the indexed points in
+    lattice, which are placed in 'h' and 'k' coordinates in the PointLists in the array.
+    If `mask` is specified, only the locations where mask is True are indexed.
+
+    Accepts:
+        pointlistarray          (PointListArray) the braggpeaks pointlist to index
+        lattice           (PointList) a PointList containing with indexed peaks,
+                                    with the four coordinates 'qx', 'qy', 'h', and 'k'
+                                    for the lattice points and corresponding indices
+        maxPeakSpacing          (float) Maximum distance from the ideal lattice points
+                                    to include a peak for indexing
+        mask                    (bool)  Boolean mask, same shape as the pointlistarray,
+                                    indicating which locations should be indexed. This
+                                    can be used to index different regions of the scan
+                                    with different lattices
+
+    Returns:
+        pointlistarray          The original pointlistarray, with new coordinates 'h' and 'k'
+                                    containing the indices of each indexable peak 
+    """
+
+    assert isinstance(pointlistarray,PointListArray)
+    assert np.all([name in pointlistarray.dtype.names for name in ('qx','qy','intensity')])
+    assert isinstance(lattice, PointList)
+    assert np.all([name in lattice.dtype.names for name in ('qx','qy','h','k')])
+
+    if mask is None:
+        mask = np.ones(pointlistarray.shape,dtype=bool)
+
+    assert np.all(mask.shape == pointlistarray.shape), 'mask must have same shape as pointlistarray'
+    assert mask.dtype == bool, 'mask must be boolean'
+
+    # add the coordinates if they don't exist
+    if not (('h' in pointlistarray.dtype.names) | ('k' in pointlistarray.dtype.names)):
+        pointlistarray = pointlistarray.add_coordinates([('h',int),('k',int)])
+
+    # loop over all the scan positions
+    for Rx in range(mask.shape[0]):
+        for Ry in range(mask.shape[1]):
+            if mask[Rx,Ry]:
+                pl = pointlistarray.get_pointlist(Rx,Ry)
+                #from pdb import set_trace
+                #set_trace()
+
+                for i in range(pl.length):
+                    r2 = (pl.data['qx'][i]-lattice.data['qx'])**2 + \
+                         (pl.data['qy'][i]-lattice.data['qy'])**2
+                    ind = np.argmin(r2)
+                    if r2[ind] <= maxPeakSpacing**2:
+                        pl.data['h'][i] = lattice.data['h'][ind]
+                        pl.data['k'][i] = lattice.data['k'][ind]
+                    else:
+                        pl.data['h'][i] = 8675309 # turns out int's can be NaN
```

Comment:
Hmm.  I'm not crazy about this approach.  Perhaps we should define an additional coord which contains a boolean mask, specifying where the index values are meaningful.  Unfortunately, I don't think we want to use np.ma here, as but simply adding a mask coord could do the trick.

Reply:
now adds `'index_mask'` coordinate